File: src/c-leetcode/first/Code_110_BalancedBinaryTree.c
/* balanced binary tree

   gcc Code_110_BalancedBinaryTree.c ../TreeNode.c
   
   @since 2020-8-27 Thursday 11:12 - 11:33
 */

#include "../TreeNode.h"
#include <stdio.h>
#include <stdbool.h>

int max(int i, int j);
int abs(int i);
int maxDepth(struct TreeNode *root);

bool isBalanced(struct TreeNode *root)
{
    if (root == 0)
        return true;
    if (abs(maxDepth(root->left) - maxDepth(root->right)) > 1)
        return false;
    else
        return isBalanced(root->left) && isBalanced(root->right);
}

int maxDepth(struct TreeNode *root)
{
    if (root == 0)
        return 1;
    else
        return max(maxDepth(root->left), maxDepth(root->right)) + 1;
}

int max(int i, int j)
{
    return i < j ? j : i;
}

int abs(int i)
{
    return i > 0 ? i : -i;
}

int main(int argc, char const *argv[])
{
    bool res = isBalanced(binaryTree());
    printf("%d\n", res);
}

File: src/leetcode/first/Code_110_BalancedBinaryTree.java
package leetcode.first;

import leetcode.tool.TreeNode;

/**
 * 平衡二叉树
 *
 * @since 2020-4-18 Saturday 12:05
 */
public class Code_110_BalancedBinaryTree {
    static boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        else return Math.abs(depth(root.left) - depth(root.right)) < 2 && isBalanced(root.left) && isBalanced(root.right);
    }

    private static int depth(TreeNode root) {
        if (root == null) return 0;
        else return Math.max(depth(root.left), depth(root.right)) + 1;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(1);
        System.out.println(isBalanced(root));
    }
}
